src/rebase.c

```diff
@@ -975,10 +977,40 @@ static int rebase_commit__create(
 		message = git_commit_message(current_commit);
 	}
 
-	if ((error = git_commit_create(&commit_id, rebase->repo, NULL, author,
-		committer, message_encoding, message, tree, 1,
-		(const git_commit **)&parent_commit)) < 0 ||
-		(error = git_commit_lookup(&commit, rebase->repo, &commit_id)) < 0)
+	/* this error will be cleared by the signing process, but should be set
+	 * to signal the unsigned commit create process if we are not going to sign */
+	error = GIT_PASSTHROUGH;
+	if (rebase->options.signature_cb) {
+		if ((error = git_commit_create_buffer(&commit_content, rebase->repo, author, committer,
+				message_encoding, message, tree, 1, (const git_commit **)&parent_commit)) < 0)
+			goto done;
+
+		if ((error = rebase->options.signature_cb(&signature, git_buf_cstr(&commit_content),
+				rebase->options.payload)) < 0 && error != GIT_PASSTHROUGH)
+			goto done;
+
+		if (error != GIT_PASSTHROUGH) {
+			if (rebase->options.signature_field_cb &&
+				(error = rebase->options.signature_field_cb(&signature_field, rebase->options.payload)) < 0) {
```

Comment:
I'm also wondering about the genericity of this : given that this is mostly about customizing the commit creation when rebasing, I'd wager we'd hit an explosion of callbacks if someone encounters #3075 (like #4667 but for commits).

Reply:
@tiennou Are you suggesting something along the lines of `git_rebase_commit_ext` or `git_rebase_commit_with_options`?

Reply:
Yeah, or a lower-level struct that would wrap all general commit-related information in it, instead of parameter-combinations to a couple of functions plus options for different behaviors.

```c
typedef struct {
    int version;
    git_tree *tree;
    git_buf *msg;
    git_oidarray parents;
    git_signature *author;
    git_signature *committer;
} git_commit_data;
```

This would make it possible to add custom commit headers (`encoding` being the only one we provide access to at the moment), or the signing callback you're after.

Maybe a more `git_treebuilder`-like API ?

I'd be glad to help with the API design, but I agree we should discuss it first 😉.

Reply:
Hmm.  One option is to have a commit creation callback - if provided, we just hand you the tree, parents, etc, and you hand us back a commit ID when you're done.

Reply:
Rereading, just to clarify : my initial suggestion was to add a lower-level `git_commit_create` variant that takes a commit-description structure (and maybe some behavioral options but that's part of what #3075 is about). Then that struct would be provided to the callback (the option @ethomson proposes), also insulating that callback's signature from unforeseen parameter changes.

Reply:
Is this a direction that I should be moving this PR in or will the additional callback added to `git_rebase_options` suffice for now?